contracts/pox4-pools.clar

```diff
@@ -0,0 +1,370 @@
+                 (user-pox-addr {hashbytes: (buff 32), version: (buff 1)})
+                 (user-metadata (optional {keys: (list 30 (string-ascii 8)), values: (list 30 (string-ascii 80))})))
+  (begin
+    ;; Must be called directly by the tx-sender or by an allowed contract-caller
+    (asserts! (check-caller-allowed) err-stacking-permission-denied)
+    (match user-metadata
+      md (map set-metadata-internal (get keys md) (get values md))
+      (list true))
+    (map-set user-data tx-sender
+      {pox-addr: user-pox-addr, cycle: (contract-call? 'ST000000000000000000002AMW42H.pox-4 current-pox-reward-cycle)})
+    (delegate-stx-inner amount-ustx delegate-to until-burn-ht)))
+
+;; @desc Pool admins call this function to lock stacks of their pool members in batches for 1 cycle.
+;; @param users; list of users with amounts to lock.
+;; @param pox-address; the pool's bitcoin reward address.
+;; @param start-burn-ht; a future bitcoin height of the current cycle.
+(define-public (delegate-stack-stx (users (list 30 {user: principal, amount-ustx: uint}))
+                 (pox-address { version: (buff 1), hashbytes: (buff 32)})
+                 (start-burn-ht uint))
+  (begin
+    (asserts! (check-caller-allowed) err-stacking-permission-denied)
+    (ok (get result
+          (fold delegate-stack-stx-fold users {start-burn-ht: start-burn-ht, pox-address: pox-address, result: (list)})))))
+
+;; @desc Pool admins call this function to lock stacks of their pool members in batches for a lock period of 1 cycle.
+;; The locking amount is determined from the delegated amount and the users balances.
+;; @param users; list of current pool members.
+;; @param pox-address; the pool's bitcoin reward address.
+;; @param start-burn-ht; a future bitcoin height of the current cycle.
+(define-public (delegate-stack-stx-simple (users (list 30 principal))
+                 (pox-address { version: (buff 1), hashbytes: (buff 32)})
+                 (start-burn-ht uint))
+  (begin
+    (asserts! (check-caller-allowed) err-stacking-permission-denied)
+    (ok (get result
+          (fold delegate-stack-stx-simple-fold users {start-burn-ht: start-burn-ht, pox-address: pox-address, result: (list)})))))
+;;
+;; Read-only functions
+;;
+
+;; Returns the user's stacking details from pox contract,
+;; the user's delegation details from "user-data" and the
+;; total locked stacks for the given pool and cycle-id.
+(define-read-only (get-status (pool principal) (user principal) (cycle-id uint))
+  (let ((stacker-info (unwrap! (pox-get-stacker-info user) err-no-stacker-info)))
+    (ok {stacker-info: stacker-info,
+         user-info: (unwrap! (map-get? user-data user) err-no-user-info),
+         total: (get-total pool cycle-id)})))
+
+;; Returns the number of lists of stackers that have locked their stx for the given pool and cycle.
+(define-read-only (get-status-lists-last-index (pool principal) (reward-cycle uint))
```

Comment:
If I understood correctly how this works, this function should be renamed to `get-status-lists-len` because it is returning a length, not an index. 

As currently named, it seems that the value returned is an index, and therefore should be usable as an argument in `get-status-list`, but it can't. For example, when accessing the "last" map entry, the index would actually be "length - 1".

I hope the above makes sense, the wording can get quite confusing. Essentially, if we think of the `grouped-stackers` map as an array of groups, I would expect the length to be the number of groups, and the index of the last group to be (length - 1).

Reply:
The naming comes from the implementation pattern of a map of lists.
We can discuss the naming. Best to add more explanations to the readme